Sources/SentryCrash/Recording/SentryCrashBinaryImageCache.c

```diff
@@ -0,0 +1,115 @@
+#include "SentryCrashBinaryImageCache.h"
+#include "SentryCrashDynamicLinker.h"
+#include <mach-o/dyld.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define CACHE_SIZE_INCREMENT 100
+static SentryCrashBinaryImage *binaryImagesBuffer = NULL;
+static uint binaryImagesBufferLength;
+static int binaryImagesAmount;
+static pthread_mutex_t binaryImagesMutex = PTHREAD_MUTEX_INITIALIZER;
+
+static void
+increaseBufferSize(void)
+{
+    pthread_mutex_lock(&binaryImagesMutex);
+    uint oldLength = binaryImagesBufferLength;
+    uint newLength = oldLength + CACHE_SIZE_INCREMENT;
+
+    SentryCrashBinaryImage *newBuffer = malloc(sizeof(SentryCrashBinaryImage) * newLength);
+    memcpy(newBuffer, binaryImagesBuffer, sizeof(SentryCrashBinaryImage) * oldLength);
+    free(binaryImagesBuffer);
+
+    binaryImagesBuffer = newBuffer;
+    binaryImagesBufferLength = newLength;
+    pthread_mutex_unlock(&binaryImagesMutex);
+}
+
+static void
+binaryImageAdded(const struct mach_header *header, intptr_t slide)
+{
+    if (binaryImagesBuffer == NULL) {
+        return;
+    }
+
+    Dl_info info;
+    if (!dladdr(header, &info) || info.dli_fname == NULL) {
+        return;
+    }
+
+    SentryCrashBinaryImage binaryImage = { 0 };
+    if (!sentrycrashdl_getBinaryImageForHeader(
+            (const void *)header, info.dli_fname, &binaryImage)) {
+        return;
+    }
+
+    if (binaryImagesAmount >= binaryImagesBufferLength) {
+        increaseBufferSize();
+    }
+
+    binaryImagesBuffer[binaryImagesAmount++] = binaryImage;
```

Comment:
Do we need to take a lock before reading `binaryImagesAmount` and modifying `binaryImagesBuffer `?
```suggestion
    pthread_mutex_lock(&binaryImagesMutex);
    if (binaryImagesAmount >= binaryImagesBufferLength) {
         increaseBufferSize();
     }
    binaryImagesBuffer[binaryImagesAmount++] = binaryImage;
    pthread_mutex_unlock(&binaryImagesMutex);
```